django/utils/http.py

```diff
@@ -303,3 +306,151 @@ def is_safe_url(url, host=None):
+        pairs = FIELDS_MATCH.split(qs)
+    r = []
+    for name_value in pairs:
+        if not name_value and not strict_parsing:
+            continue
+        nv = name_value.split(b'=', 1)
+        if len(nv) != 2:
+            if strict_parsing:
+                raise ValueError("bad query field: %r" % (name_value,))
+            # Handle case of a control-name with no equal sign
+            if keep_blank_values:
+                nv.append(b'')
+            else:
+                continue
+        if len(nv[1]) or keep_blank_values:
+            name = unquote(nv[0].replace(b'+', b' '))
+            value = unquote(nv[1].replace(b'+', b' '))
+            r.append((name, value))
+
+    return r
+
+
+_implicit_encoding = 'ascii'
+_implicit_errors = 'strict'
+
+
+def _noop(obj):
+    return obj
+
+
+def _encode_result(obj, encoding=_implicit_encoding, errors=_implicit_errors):
+    return obj.encode(encoding, errors)
+
+
+def _decode_args(args, encoding=_implicit_encoding, errors=_implicit_errors):
+    return tuple(x.decode(encoding, errors) if x else '' for x in args)
+
+
+def _coerce_args(*args):
+    # Invokes decode if necessary to create str args
+    # and returns the coerced inputs along with
+    # an appropriate result coercion function
+    #   - noop for str inputs
+    #   - encoding function otherwise
+    str_input = isinstance(args[0], str)
+    for arg in args[1:]:
+        # We special-case the empty string to support the
+        # "scheme=''" default argument to some functions
+        if arg and isinstance(arg, str) != str_input:
+            raise TypeError("Cannot mix str and non-str arguments")
+    if str_input:
+        return args + (_noop,)
+    return _decode_args(args) + (_encode_result,)
+
+
+def limited_parse_qsl_py3(qs, keep_blank_values=False, strict_parsing=False,
+               encoding='utf-8', errors='replace', fields_limit=None):
+    """
+    Parse a query given as a string argument and return a list.
+
+    Lifted from urlparse with an additional "fields_limit" argument.
+
+    Arguments:
+
+    qs: percent-encoded query string to be parsed
+
+    keep_blank_values: flag indicating whether blank values in
+        percent-encoded queries should be treated as blank strings.  A
+        true value indicates that blanks should be retained as blank
+        strings.  The default false value indicates that blank values
+        are to be ignored and treated as if they were  not included.
+
+    strict_parsing: flag indicating what to do with parsing errors. If
+        false (the default), errors are silently ignored. If true,
+        errors raise a ValueError exception.
+
+    encoding and errors: specify how to decode percent-encoded sequences
+        into Unicode characters, as accepted by the bytes.decode() method.
+
+    fields_limit: maximum number of fields parsed or an exception
+        is raised. None means no limit and is the default.
+    """
+    qs, _coerce_result = _coerce_args(qs)
```

Comment:
When we call this function, we already ensured that we have strings (https://github.com/edevil/django/blob/21231/django/http/request.py#L391-L397) -- I do not think there is need for this coercing step, this allows you to drop _coerce_result too and should make it easier to merge the py2 and py3 function.